File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.stream.Collectors;

public class Anagram {
  String s;

  Anagram(String s) {
    this.s = s;
  }

  String generateKey(String s) {
    return s.toLowerCase()
        .chars()
        .sorted()
        .mapToObj(c -> (char) c)
        .map(String::valueOf)
        .collect(Collectors.joining());
  }

  List<String> match(List<String> candidates) {
    return candidates.stream()
        .filter(
            candidate ->
                !candidate.equalsIgnoreCase(s) && generateKey(candidate).equals(generateKey(s)))
        .toList();
  }
}

File: cpp/anagram/anagram.cpp
#include "anagram.h"

namespace anagram
{
anagram::anagram(const std::string &word) : word{word} {}

std::string str_tolower(std::string s)
{
    std::transform(s.begin(), s.end(), s.begin(), [](char ch) { return std::tolower(ch); });

    return s;
}

std::string generate_key(std::string s)
{
    s = str_tolower(s);
    std::sort(s.begin(), s.end());

    return s;
}

std::vector<std::string> anagram::matches(const std::vector<std::string> &candidates)
{
    std::vector<std::string> result;
    for (const auto &candidate : candidates)
    {
        if (str_tolower(candidate) != str_tolower(word) && generate_key(candidate) == generate_key(word))
        {
            result.push_back(candidate);
        }
    }

    return result;
}
} // namespace anagram
